### ISSUE TITLE
USB scanner not working when TCP/IP SYN cookies is disabled

### ISSUE THREAD

**Issue by ghost** [utsushi/utsushi#47]
When TCP/IP SYN cookies is disabled, utsushi is able to detect my scanner but it "cannot connect":
```
$ scanimage -L
2018-Feb-24 14:38:29.975467[139664597085184]: sane_utsushi_init: SANE 'utsushi' backend (1.1.0), a part of Utsushi 0.36.0
2018-Feb-24 14:38:29.975558[139664597085184]: sane_utsushi_get_devices: invalidated SANE_Device pointers
2018-Feb-24 14:38:30.150787[139664597085184]: some device not found. delete combo::?adf=esci:usb:04b8:0155&fb=esci:gt-s650:usb:04b8:013c
2018-Feb-24 14:38:30.151233[139664597085184]: some device not found. delete combo::?adf=esci:usb:04b8:0156&fb=esci:gt-s650:usb:04b8:013c
2018-Feb-24 14:38:30.151633[139664597085184]: some device not found. delete combo::?adf=esci:usb:04b8:0157&fb=esci:gt-s650:usb:04b8:013c
2018-Feb-24 14:38:30.152128[139664597085184]: some device not found. delete combo::?adf=esci:usb:04b8:0155&fb=esci:gt-s650:usb:04b8:013d
2018-Feb-24 14:38:30.152527[139664597085184]: some device not found. delete combo::?adf=esci:usb:04b8:0156&fb=esci:gt-s650:usb:04b8:013d
2018-Feb-24 14:38:30.152938[139664597085184]: some device not found. delete combo::?adf=esci:usb:04b8:0157&fb=esci:gt-s650:usb:04b8:013d
2018-Feb-24 14:38:30.152966[139664597085184]: sane_utsushi_get_devices: added esci:gt-s650:usb:/sys/devices/pci0000:00/0000:00:1d.7/usb2/2-2/2-2:1.0 to device pool
device `utsushi:esci:gt-s650:usb:/sys/devices/pci0000:00/0000:00:1d.7/usb2/2-2/2-2:1.0' is a EPSON Epson_Perfection_V39 
2018-Feb-24 14:38:30.156378[139664597085184]: sane_utsushi_exit: closing open handles
```
```
$ scanimage
2018-Feb-24 14:36:57.408078[139970900331520]: sane_utsushi_init: SANE 'utsushi' backend (1.1.0), a part of Utsushi 0.36.0
2018-Feb-24 14:36:57.408169[139970900331520]: sane_utsushi_get_devices: invalidated SANE_Device pointers
2018-Feb-24 14:36:57.587357[139970900331520]: some device not found. delete combo::?adf=esci:usb:04b8:0155&fb=esci:gt-s650:usb:04b8:013c
2018-Feb-24 14:36:57.587803[139970900331520]: some device not found. delete combo::?adf=esci:usb:04b8:0156&fb=esci:gt-s650:usb:04b8:013c
2018-Feb-24 14:36:57.588378[139970900331520]: some device not found. delete combo::?adf=esci:usb:04b8:0157&fb=esci:gt-s650:usb:04b8:013c
2018-Feb-24 14:36:57.588776[139970900331520]: some device not found. delete combo::?adf=esci:usb:04b8:0155&fb=esci:gt-s650:usb:04b8:013d
2018-Feb-24 14:36:57.589199[139970900331520]: some device not found. delete combo::?adf=esci:usb:04b8:0156&fb=esci:gt-s650:usb:04b8:013d
2018-Feb-24 14:36:57.589596[139970900331520]: some device not found. delete combo::?adf=esci:usb:04b8:0157&fb=esci:gt-s650:usb:04b8:013d
2018-Feb-24 14:36:57.589623[139970900331520]: sane_utsushi_get_devices: added esci:gt-s650:usb:/sys/devices/pci0000:00/0000:00:1c.4/0000:0d:00.0/usb9/9-1/9-1:1.0 to device pool
2018-Feb-24 14:36:57.589770[139970900331520]: sane_utsushi_open: creating SANE_Handle for esci:gt-s650:usb:/sys/devices/pci0000:00/0000:00:1c.4/0000:0d:00.0/usb9/9-1/9-1:1.0
2018-Feb-24 14:36:57.589799[139970900331520]: looking for preloaded 'esci' driver
2018-Feb-24 14:36:57.589831[139970900331520]: looking for 'esci' driver in '/usr/lib/utsushi'
2018-Feb-24 14:36:57.593391[139970900331520]: using '/usr/lib/utsushi/libdrv-esci'
2018-Feb-24 14:36:57.594237[139970900331520]: /usr/lib/utsushi/esci-gt-s650[6295]: starting
2018-Feb-24 14:37:00.612926[139970900331520]: connect: Operation now in progress
2018-Feb-24 14:37:04.643953[139970900331520]: connect: Operation now in progress
2018-Feb-24 14:37:08.675951[139970900331520]: connect: Operation now in progress
2018-Feb-24 14:37:12.707941[139970900331520]: connect: Operation now in progress
2018-Feb-24 14:37:16.739965[139970900331520]: connect: Operation now in progress
2018-Feb-24 14:37:16.740197[139970543068928]: terminating /usr/lib/utsushi/esci-gt-s650 (port 51319)
2018-Feb-24 14:37:16.740330[139970900331520]: sane_utsushi_open: unhandled exception
usb:/sys/devices/pci0000:00/0000:00:1c.4/0000:0d:00.0/usb9/9-1/9-1:1.0: cannot connect
2018-Feb-24 14:37:16.740365[139970900331520]: sane_utsushi_open: Operation not supported
scanimage: open of device utsushi:esci:gt-s650:usb:/sys/devices/pci0000:00/0000:00:1c.4/0000:0d:00.0/usb9/9-1/9-1:1.0 failed: Operation not supported
2018-Feb-24 14:37:16.742691[139970900331520]: sane_utsushi_exit: closing open handles
```
```
$ dmesg
TCP: request_sock_TCP: Possible SYN flooding on port 18149. Dropping request.  Check SNMP counters.
TCP: request_sock_TCP: Possible SYN flooding on port 56431. Dropping request.  Check SNMP counters.
```
Adjusting the values of `/proc/sys/net/core/somaxconn` and `/proc/sys/net/ipv4/tcp_max_syn_backlog` to over 130000 does not solve the problem.
Why is this happening despite I am using a USB scanner? I noticed that in iscan (V2), I do not have to enable TCP/IP SYN cookies for my scanner to work.

**Comment by paddy-hack** [comment]
The V39 requires a non-free component.  With Utsushi and Image Scan v3, this is done via IPC where the non-free component is running as a separate process, hence the need for TCP/IP.  With Image Scan! for Linux (V2, a *completely* different beast) this works via a dynamically loaded libary.

I suggest you use your V39 with Image Scan! for Linux if that works for you.  If you have any specific reason to use Image Scan v3, feel free to do so but I have *no* intent of debugging issues with EPSON's **non-free** components.  Please use whatever support contact EPSON gives for their downloadable drivers.

If it is just a plain timeout issue though, you may want to play with the `tries_left` value in `lib/connexion.cpp` and the timeout value passed to `set_timeout()` in `connexion::connect_()` in the same file.

BTW, thanks for making me aware of 3.36.0.  Looks like I need to sync :wink:

**Comment by gojpdchx** [comment]
@paddy-hack 
Thank you for the information. Is `utsushi` developed by both EPSON and non-EPSON developers?

The non-free V39 plugin I installed comes from `https://download2.ebz.epson.net/imagescanv3/fedora/latest1/rpm/x64/imagescan-bundle-fedora-27-1.3.26.x64.rpm.tar.gz`.

The `utsushi` I used is compiled from `http://support.epson.net/linux/src/scanner/imagescanv3/fedora/imagescan-3.36.0-1epson4fedora27.src.rpm`. **The version is labelled as "3.36.0" but it does _not_ include the patch to fix #41.**

**Comment by paddy-hack** [comment]
I used to be heavily involved in the development of what is distributed from the EPSON Download Center.  Since 2016-04 I am no longer involved at all.  I started the utsushi/imagescan and utsushi/utsushi projects to provide a public record of changes to the "upstream" sources and to scratch the occasional personal itch.  I still track changes to the upstream sources but I haven't really had any itches of late that I felt needed scratching.

I haven't looked at 3.36.0 yet (probably just adds support for more models), but you can build from the latest sources of this project and should be able to use the non-free plugin you got from the EPSON Download Center.

**Comment by paddy-hack** [comment]
I've just merged the changes from imagescan-3.36.0 to the utsushi master branch in 1b76c36.  As expected it only adds support for a few new models.  So if you really want that source code base + a fix for #41 (and some more), it's at your fingertips.

Does that solve your issue?  If so, please close this issue.
